File: shortest-path-algorithms/johnson-algorithm/javascript/JohnsonsAlgo.js
const graph = [[0, 4, 0, 0, 1],
            [0, 0, 0, 0, 0],
            [0, 7, 0, -2, 0],
            [0, 1, 0, 0, 0],
            [0, 0, 0, -5, 0]]

allPairShortestPath(graph)


function allPairShortestPath(graph){
    const edges = []

    for(let u = 0; u < graph.length; u++){
        for(let v = 0; v < graph.length; v++){
            if( graph[u][v] != 0){
                edges.push([u, v, graph[u][v]])
            }
        }
    }

    const distance = bellmanFord(edges, graph.length)

    // new positive weights to the edges
    const newGraph = []

    for(let i = 0; i < graph.length; i++){
        newGraph.push(new Array(graph.length).fill(0))
    }

    for(let u = 0; u < graph.length; u++){
        for(let v = 0; v < graph.length; v++){
            if( graph[u][v] != 0){
                newGraph[u][v] = distance[u] + graph[u][v] - distance[v]
            }
        }
    }

    console.log("New Graph : ", newGraph)

    //Run dijkstra for every vertex as source
    for(let src = 0; src < graph.length; src++){
        console.log("Shortest distances from vertex "+ src)
        dijkstra(graph, newGraph, src)
    }
}

function dijkstra(oldGraph, newGraph, src){

    const v = newGraph.length
    const visited = new Array(v).fill(false)
    const distance = new Array(v).fill(Number.MAX_SAFE_INTEGER)

    distance[src] = 0

    for(let i = 0; i < v -1 ; i++){
        minVertex = findMinVertex(distance, visited)
        visited[minVertex] = true

        for(let j = 0; j < v; j++){
            if(oldGraph[minVertex][j] != 0 && !visited[j]){
                let newDist = distance[minVertex] + newGraph[minVertex][j]

                if(newDist < distance[j]){
                    distance[j] = newDist
                }
            }
        }
    }

    console.log(distance)
}

function findMinVertex(distance, visited){
    let minVertex = -1

    for(let i = 0; i < distance.length; i++){
        if(!visited[i] && (minVertex == -1 || distance[i] < distance[minVertex]))
            minVertex = i;
    }

    return minVertex
}

function bellmanFord(edges, V){
    //arbitrary source
    const dist = new Array(V+1).fill(Number.MAX_SAFE_INTEGER)

    dist[V] = 0

    //add edges from source to all other vertices
    for(let i = 0; i < V; i++){
        edges.push([V, i, 0])
    }

    for(let i = 0; i < V; i++){
    edges.forEach(
        li => {
            let u = li[0]
            let v = li[1]
            let w = li[2]

            if( dist[u] != Number.MAX_SAFE_INTEGER && dist[u] + w < dist[v]){
                dist[v] = dist[u] + w
             }
            }
        )   
    }

    //discard the arbitrary source
    return dist.splice(0, V)
}

File: shortest-path-algorithms/johnson-algorithm/python/JohnsonsAlgo.py

def allPairShortestPath(graph):
    edges = []

    for u in range(len(graph)):
        for v in range(len(graph)):
            if graph[u][v] != 0:
                edges.append([u, v, graph[u][v]])

    distance = bellman_ford(edges, len(graph))

    # new positive edge weights
    new_graph = [[0 for x in range(len(graph))] for y in range(len(graph))]

    for u in range(len(graph)):
        for v in range(len(graph)):
            if graph[u][v] != 0:
                new_graph[u][v] = distance[u] + graph[u][v] - distance[v]

    print('New graph : ' + str(new_graph))

    # run dijkstra for every vertex as source
    for src in range(len(graph)):
        print('shortest distance from vertex : ' , src)
        dijkstra(graph, new_graph, src)


def dijkstra(old_graph, new_graph, src):
    v = len(new_graph)
    visited = [False for i in range(v)]
    distance = [float('Inf')] * v

    distance[src] = 0

    for i in range(v - 1):
        min_vertex = find_min_vertex(distance, visited)
        visited[min_vertex] = True

        for j in range(v):
            if old_graph[min_vertex][j] != 0 and not visited[j]:
                new_dist = distance[min_vertex] + new_graph[min_vertex][j]
                if new_dist < distance[j]:
                    distance[j] = new_dist

    for i in range(v):
        print(i, ' ', distance[i])


def find_min_vertex(distance, visited):
    min_vertex = -1

    for i in range(len(distance)):
        if (min_vertex == -1 or distance[min_vertex] > distance[i]) and not visited[i]:
            min_vertex = i

    return min_vertex


def bellman_ford(edges, vertices):
    # arbitrary source
    dist = [float('Inf')] * (vertices + 1)

    dist[vertices] = 0

    # added edges to all the other vertices from source
    for i in range(vertices):
        edges.append([vertices, i, 0])

        # vertices -1
    for i in range(vertices):
        for (u, v, w) in edges:
            if dist[u] != float('Inf') and (dist[u] + w < dist[v]):
                dist[v] = dist[u] + w

    # discard the arbitrary source
    return dist[0: vertices]


graph = [[0, 4, 0, 0, 1],
         [0, 0, 0, 0, 0],
         [0, 7, 0, -2, 0],
         [0, 1, 0, 0, 0],
         [0, 0, 0, -5, 0]]

# Johnson's algo
allPairShortestPath(graph)
